core/src/main/java/org/apache/iceberg/encryption/EnvelopeEncryptionManager.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.encryption;
+
+import java.nio.ByteBuffer;
+import java.security.SecureRandom;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * Encryption manager that performs envelope encryption.
+ */
+public class EnvelopeEncryptionManager implements EncryptionManager {
+  private final EnvelopeConfig dataEncryptionConfig;
+  private final boolean pushdown;
+  private final KmsClient kmsClient;
+  private final int dataKeyLength;
+
+  private transient volatile SecureRandom workerRNG = null;
```

Comment:
Not a big deal but one thing we could do here if it makes sense is initialize and reinit this code in the serialize rather than the null check below

```
    private void readObject(java.io.ObjectInputStream in)
            throws IOException, ClassNotFoundException {
        in.defaultReadObject();
        this.workerRNG = new SecureRandom();
    }
    ```

Reply:
this field is transient, not passed in serialization. It is used locally only, and could have been a fully local var, defined and init each time a function is called. But since creating a secure random gen can be expensive, it is defined as a class member and kept for a while